client/mifare/mifarehost.c

```diff
@@ -413,27 +435,59 @@ int mfnested(uint8_t blockNo, uint8_t keyType, uint8_t *key, uint8_t trgBlockNo,
 	// the statelists now contain possible keys. The key we are searching for must be in the
 	// intersection of both lists. Create the intersection:
 	qsort(statelists[0].head.keyhead, statelists[0].len, sizeof(uint64_t), compare_uint64);
-	qsort(statelists[1].head.keyhead, statelists[1].len, sizeof(uint64_t), compare_uint64);
-	statelists[0].len = intersection(statelists[0].head.keyhead, statelists[1].head.keyhead);
+
+	if (num_unique_nonces > 1) {
+		qsort(statelists[1].head.keyhead, statelists[1].len, sizeof(uint64_t), compare_uint64);
+		statelists[0].len = intersection(statelists[0].head.keyhead, statelists[1].head.keyhead);
+	}
+	else {
+		PrintAndLog("Nonce 1 and 2 are the same!");
+	}
+
+	if (statelists[0].len > 100) {
+		PrintAndLog("We have %d keys to check. This will take a very long time!", statelists[0].len);
```

Comment:
can we estimate the time? E.g. measuring the time for the first keyblock and then multiplying with half of the remaining keyblocks?

Reply:
Unfortunately we cannot because we don't know how many keys until the correct one.  We can calculate the worst-case time, however it is very rare for it to actually take anywhere near that long.

Reply:
We can tell the expected (average) time. Statistically we will find a key after testing 50% of the candidates. And we can tell the worst case time.

Reply:
Worst-case time remaining added to the X keys left message.